File: iota-c++/tangle.cpp
#include <iostream>
#include <vector>
#include <set>
#include <ctime>
#include <chrono>
#include <iomanip>
#include <sstream>
#include <random>
#include <openssl/sha.h>
#include "./httplib.h"
#include "single_include/nlohmann/json.hpp"

namespace json = nlohmann;

class Tangle {
public:
    std::vector<json::json> nodes;
    std::set<std::string> peers;

    bool valid_proof(int last_proof, const std::string& last_hash, int proof) {
        std::string guess = std::to_string(last_proof) + last_hash + std::to_string(proof);
        std::string guess_hash = sha256(guess);
        return guess_hash.substr(0, 4) == "0000";
    }

    int proof_of_work(int last_proof, const std::string& last_hash) {
        int proof = 0;
        while (!valid_proof(last_proof, last_hash, proof)) {
            proof++;
        }
        return proof;
    }

    std::string sha256(const std::string& input) {
        unsigned char hash[SHA256_DIGEST_LENGTH];
        SHA256_CTX sha256;
        SHA256_Init(&sha256);
        SHA256_Update(&sha256, input.c_str(), input.length());
        SHA256_Final(hash, &sha256);
        std::stringstream ss;
        for (int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
            ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
        }
        return ss.str();
    }

    void validate_node(json::json& node) {
        if (nodes[node["index"]]["validity"] < 1) {
            int last_proof = nodes[node["index"]]["proof"];
            std::string last_hash;
            for (const auto& prev_hash : nodes[node["index"]]["previous_hashs"]) {
                last_hash += prev_hash;
            }
            nodes[node["index"]]["proof"] = proof_of_work(last_proof, last_hash);
            nodes[node["index"]]["validity"] += 1;
        }
    }

    json::json create_node(const json::json& data, const std::vector<int>& prev_nodes, int new_index, int validity = 0) {
        std::vector<std::string> prev_hashes;

        for (int i : prev_nodes) {
            prev_hashes.push_back(sha256(nodes[i].dump()));
            nodes[i]["next_nodes"].push_back(new_index);
        }

        json::json node = {
            {"index", new_index},
            {"timestamp", std::time(0)},
            {"data", data},
            {"proof", 0},
            {"previous_hashs", prev_hashes},
            {"previous_nodes", prev_nodes},
            {"next_nodes", json::json::array()},
            {"validity", validity},
        };

        return node;
    }

    int send_transaction(const json::json& data) {
        std::vector<json::json> node_to_attach;
        std::vector<int> nodes_indexes;
        int new_index = nodes.size();
        std::vector<json::json> worst_cases_scenario;
        std::vector<int> worst_cases_scenario_indexes;

        for (int i = nodes.size() - 1; i >= 0; --i) {
            auto& node = nodes[i];
            if (node["validity"] < 1) {
                node_to_attach.push_back(node);
                nodes_indexes.push_back(node["index"]);
            } else {
                if (worst_cases_scenario.empty() || worst_cases_scenario.size() < 2) {
                    worst_cases_scenario.push_back(node);
                    worst_cases_scenario_indexes.push_back(node["index"]);
                }
            }
            if (node_to_attach.size() == 2) {
                break;
            }
        }

        while (node_to_attach.size() < 2) {
            node_to_attach.push_back(worst_cases_scenario.back());
            nodes_indexes.push_back(worst_cases_scenario_indexes.back());
            worst_cases_scenario.pop_back();
            worst_cases_scenario_indexes.pop_back();
        }

        for (auto& node : node_to_attach) {
            validate_node(node);
        }

        nodes.push_back(create_node(data, nodes_indexes, new_index));
        return new_index;
    }

    bool valid_tangle(const json::json& tangle) {
        for (const auto& node : tangle) {
            if (node["index"] >= 1) {
                int validity_of_node = node["validity"];
                auto next_nodes = node["next_nodes"];
                if (validity_of_node < next_nodes.size()) {
                    return false;
                }
                for (int n_node : next_nodes) {
                    if (node["index"] != nodes[n_node]["previous_nodes"]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

public:
    Tangle() {
        for (int i = 0; i < 2; ++i) {
            nodes.push_back(create_node(nullptr, {}, nodes.size(), 1));
        }
    }

    bool resolve_conflicts() {
        std::set<std::string> neighbours = peers;
        json::json new_tangle = nullptr;
        size_t max_length = nodes.size();

        for (const auto& peer : neighbours) {
            httplib::Client client(peer.c_str());
            auto res = client.Get("/tangle");
            if (res && res->status == 200) {
                size_t length = res->get_header_value<size_t>("Content-Length").value_or(0);
                auto tangle = json::json::parse(res->body);

                if (length > max_length && valid_tangle(tangle)) {
                    max_length = length;
                    new_tangle = tangle;
                }
            }
        }

        if (new_tangle != nullptr) {
            nodes = new_tangle;
            return true;
        }
        return false;
    }

    void register_peer(const std::string& address) {
        peers.insert(address);
    }
};

File: tangle.py
import props as properties
import json
from time import time
import hashlib
import requests
from urllib.parse import urlparse

"""
Los nodos en la red Tangle (DAG) son representados en el formato JSON descrito:
ellos pueden ser transmitidos a través de la red
 Node = {
            'index': "índice",
            'timestamp': time(),
            'data': "datos de la transacción donde los datos se guardarán",
            'proof': "prueba de trabajo",
            'previous_hashs': "Los hashes de los dos últimos nodos conectados",
            'previous_nodes': 'índice de nodos previos',
            'next_nodes': 'índices de los nodos que apuntan a él,
            'validity': número de veces que el nodo ha sido probado
        }
"""

class Tangle(object):
	"""docstring for Tangle"""
	def __init__(self):
		super(Tangle, self).__init__()
		self.nodes = []
		self.peers = set()
		# Crea un bloque inicial
		for i in range(properties.REQUIRED_PROOF):
			self.nodes.append(self.create_node(None, [], len(self.nodes), 
			validity = properties.REQUIRED_PROOF))

		""" Se asegura que el nodo tenga el número correcto de ceros.

		Returns:
			bool: Si la cantida de ceros es correcta.  
		"""
	@staticmethod	
	def valid_proof(last_proof, last_hash, proof):
		guess = (str(last_proof) + str(last_hash) + str(proof)).encode()
		guess_hash = hashlib.sha256(guess).hexdigest()
		return guess_hash[:4] == "0000"

		"""Se hace la prueba de trabajo
		"""
	def proof_of_work(self, last_proof, last_hash):
		proof = 0
		while self.valid_proof(last_proof, last_hash, proof) is False:
			proof += 1
		return proof
		
		""" Crear n nuevo hash para un nodo 

		Returns:
			sha256: hash sha-256 del nodo
		"""
	@staticmethod
	def hash(node):
		# Hace un hash para el block
		# Nos aseguramos que el dictionario este ordenado de lo contrario tendremos hashes inconsistentes
		node_string = json.dumps(node, sort_keys=True).encode()
		return hashlib.sha256(node_string).hexdigest()

		""" Valida el nodo
		"""
	def validate_node(self, node):
		if self.nodes[node['index']]['validity'] < properties.REQUIRED_PROOF:
			last_proof = self.nodes[node['index']]['proof'] # This nodes proof
			last_hash = ""
			print(node['index'])
			for prev_hash in self.nodes[node['index']]['previous_hashs']:
				last_hash += prev_hash # the hashes of the nodes his node connects
			self.nodes[node['index']]['proof'] = self.proof_of_work(last_proof, last_hash)
			self.nodes[node['index']]['validity'] =+ 1

	
		"""Crear un nuevo nodo
		"""
	def create_node(self, data, prev_nodes, new_index, validity=0): # Estos prev_nodes son los índices en el dag que apuntan al nodo.
		prev_hashes = []
		'''
		Se podría agregar una futura actualización para
  		que actualize cada nodo para que actualize aquellos 
    	nodos que apuntes a esta transacción.
		'''

		for i in prev_nodes:
			prev_hashes.append(self.hash(self.nodes[i]))
			#Ahora les decimos a los nodos a los que estamos apuntando que les estamos apuntando   			
			self.nodes[i]['next_nodes'].append(new_index)

		Node = {
			'index': new_index,
			'timestamp': time(),
			'data': data,
			'proof': 0,
			'previous_hashs': prev_hashes,
			'previous_nodes': prev_nodes,
			'next_nodes': [],
			'validity': validity,
		}
		return Node

		""" Enviar una transacción
		"""
	def send_transaction(self, data):
		# Encontrar dos nodos en la red que esten provados. 
		node_to_attach = []
		nodes_indexes = []
		new_index = len(self.nodes)
		worst_cases_scinario = []
		worst_cases_scinario_indexes = []

		'''
		Se podría agregar una futura actualización donde esta función haga la búsqueda de forma aleatoria. 
		'''

		for i in range(len(self.nodes)-1, -1, -1):
			node=self.nodes[i]
			if node['validity'] < properties.REQUIRED_PROOF:
				node_to_attach.append(node)
				nodes_indexes.append(node['index'])
			else:
				if worst_cases_scinario == [] or len(worst_cases_scinario) < properties.NUMBER_OF_VALIDATION_NODES:
					worst_cases_scinario.append(node)
					worst_cases_scinario_indexes.append(node['index'])
			if len(node_to_attach) == properties.NUMBER_OF_VALIDATION_NODES:
				break

		# Si no hay suficientes transacciones varificadas, use transacciones verificadas

		while len(node_to_attach) < properties.NUMBER_OF_VALIDATION_NODES:
			node_to_attach.append(worst_cases_scinario.pop())
			nodes_indexes.append(worst_cases_scinario_indexes.pop())
		
		# Nodo en proceso de adjuntar
		for node in node_to_attach:
			self.validate_node(node)
   
		# Se adjunta el nodo a la red Tangle(DAG)
		self.nodes.append(self.create_node(data, nodes_indexes, new_index))
		return new_index


		"""Este es el algoritmo de consenso
		"""
	def valid_tangle(self, tangle):
		for node in tangle:
			if node['index'] >= properties.NUMBER_OF_VALIDATION_NODES: # do not test genesis nodes
				validity_of_node = node['validity']
				# Se asegura que el mismo número de nodos que posee. 
				# Se valida el nodo y su nivel de validad. 
				next_nodes = node['next_nodes']
				if validity_of_node < len(next_nodes):
					return False
				# Se asegura que esos nodos estan apuntado al él.
				for n_node in next_nodes:
					if node['index'] not in tangle[n_node]['previous_nodes']:
						return False
		return True

		"""Se registra un peer
		"""
	def register_peer(self, address):
		parsed_url = urlparse(address)
		self.peers.add(parsed_url.netloc)

		""" Algoritmo que verifica si este algún conflicto con los nodos
		"""
	def resolve_conflicts(self):
		neighbours = self.peers
		new_tangle = None

		# Solo buscamos una cadena que larga que la que se está procesando. 
		max_length = len(self.nodes)

		# Obtiene y verifica las cadenas de todos los pares de nuestra red.
		for peer in neighbours:
			response = requests.get("http://"+ str(peer) +"/tangle")
			if response.status_code == 200:
				length = response.json()['length']
				tangle = response.json()['tangle']

				# Compruebe si la longitud es más larga y la cadena es válida
				if length > max_length and self.valid_tangle(tangle):
					max_length = length
					new_tangle = tangle

		# Reemplazar nuestro chan si descubrimos una cadena nueva y válida más larga que la nuestra
		if new_tangle:
			self.nodes = new_tangle
			return True
		return False		
